### PR TITLE
Function unused parameter part 2

### PR DESCRIPTION
**Pull request by hrkrshnn**
Part 2 of https://github.com/ethereum/solidity/pull/9397 Rewrites functions whose return values are not used at callsite.

### CODE REVIEW

**Comment by stackenbotten** [comment]
There was an error when running `chk_coding_style` for commit `c251a570393c4d768a6fbba5fa8bfd0f401495fb`:
```
Coding style error:
 libyul/optimiser/UnusedFunctionParameterPruner.cpp:92: private:

```
Please check that your changes are working as intended.

File: libyul/optimiser/UnusedFunctionParameterPruner.cpp [hunk 1]
```diff
@@ -0,0 +1,409 @@
...
+		m_references(_references)
+
+	{
+		(*this)(_block);
+	}
+	using ASTWalker::operator();
+	void operator()(VariableDeclaration const& _v) override;
+
+	map<YulString, vector<bool>> m_returns;
+ private:
```
Comment (stackenbotten) [inline libyul/optimiser/UnusedFunctionParameterPruner.cpp:92]: Coding style error

File: libyul/optimiser/UnusedFunctionParameterPruner.cpp [hunk 2]
```diff
@@ -64,7 +64,60 @@ vector<T> applyBooleanMask(vector<T> const& _vec, vector<bool> const& _mask)
...
```
Comment (stackenbotten) [inline libyul/optimiser/UnusedFunctionParameterPruner.cpp:92]: Coding style error

**Comment by stackenbotten** [comment]
There was an error when running `chk_coding_style` for commit `50e64c06388173a5c553547c2f2936ccaa4340d8`:
```
Coding style error:
 libyul/optimiser/UnusedFunctionParameterPruner.cpp:92: private:

```
Please check that your changes are working as intended.

**Comment by stackenbotten** [comment]
There was an error when running `chk_coding_style` for commit `b9a75a73313c61ed23169fe11cad1d0c6e6ce201`:
```
Coding style error:
 libyul/optimiser/UnusedFunctionParameterPruner.cpp:92: private:

```
Please check that your changes are working as intended.

File: libyul/optimiser/UnusedFunctionParameterPruner.cpp [hunk 3]
```diff
@@ -64,7 +64,60 @@ vector<T> applyBooleanMask(vector<T> const& _vec, vector<bool> const& _mask)
...
+		m_references(_references)
+
+	{
+		(*this)(_block);
+	}
+	using ASTWalker::operator();
+	void operator()(VariableDeclaration const& _v) override;
+
+	map<YulString, vector<bool>> callSiteMasks;
+ private:
```
Comment (stackenbotten) [inline libyul/optimiser/UnusedFunctionParameterPruner.cpp:92]: Coding style error

**Comment by stackenbotten** [comment]
There was an error when running `chk_coding_style` for commit `ffd478868ca1be5c59a056c37ec5d6a0d82271ed`:
```
Coding style error:
 libyul/optimiser/UnusedFunctionParameterPruner.cpp:94: private:

```
Please check that your changes are working as intended.

File: libyul/optimiser/UnusedFunctionParameterPruner.cpp [hunk 4]
```diff
@@ -64,7 +64,63 @@ vector<T> applyBooleanMask(vector<T> const& _vec, vector<bool> const& _mask)
...
+		(*this)(_block);
+	}
+	using ASTWalker::operator();
+	void operator()(VariableDeclaration const& _v) override;
+
+	/// Function name and a boolean mask, where `false` at index `i` indicates that the function
+	/// return-parameter at index `i` in `FunctionDefinition::parameter` is unused at every function
+	/// call-site.
+	map<YulString, vector<bool>> callSiteMasks;
+ private:
```
Comment (stackenbotten) [inline libyul/optimiser/UnusedFunctionParameterPruner.cpp:94]: Coding style error

File: libyul/optimiser/UnusedFunctionParameterPruner.h [hunk 5]
```diff
@@ -34,11 +34,21 @@ namespace solidity::yul
...
  *
- * The step LiteralRematerialiser is not required for correctness. It helps deal with cases such as:
- * `function f(x) -> y { revert(y, y} }` where the literal `y` will be replaced by its value `0`,
- * allowing us to rewrite the function.
+ * The steps LiteralRematerialiser and SSATransform are not required for correctness.
+
+ * LiteralRematerialiser helps deal with cases such as: `function f(x) -> y { revert(y, y} }` where
+ * the literal `y` will be replaced by its value `0`, allowing us to rewrite the function.
+
+ * The property "return variable is unused" is inferred by counting references to return variables.
```
Comment (chriseth) [inline libyul/optimiser/UnusedFunctionParameterPruner.h:45]: Please make it clear whether the return variable is unused in the function or at the call site.

**Comment by hrkrshnn** [comment]
Replaced by https://github.com/ethereum/solidity/pull/10080

**Pull request closed without merging**